packages/kbn-optimizer/src/worker/webpack.config.ts

```diff
@@ -94,10 +94,13 @@ export function getWebpackConfig(bundle: Bundle, bundleRefs: BundleRefs, worker:
     module: {
       // no parse rules for a few known large packages which have no require() statements
       // or which have require() statements that should be ignored because the file is
-      // already bundled with all its necessary depedencies
+      // already bundled with all its necessary dependencies
       noParse: [
         /[\/\\]node_modules[\/\\]lodash[\/\\]index\.js$/,
         /[\/\\]node_modules[\/\\]vega[\/\\]build[\/\\]vega\.js$/,
+
+        // Ignore @elastic/analytics server-side shipper implementation:
+        /[\/\\]node_modules[\/\\]@elastic[\/\\]analytics[\/\\]target_node[\/\\]shippers[\/\\]elastic_v3[\/\\]server_shipper[\/\\]index\.js$/,
```

Comment:
I think we might need to split this package since it seems to have some server-specific code and some ui-specific code. We should probably have `@kbn/analytics-client` and `@kbn/analytics-server` or something. Common code could go into `@kbn/analytics-common`.

Reply:
FWIW, the client is the same for both, the UI and the Server. The difference is the implementation of the transport layer that ships the events to our Remote Telemetry Service.

There is a piece of logic deciding which implementation to import and use depending on the environment:

```
if (isBrowser()) {
  import('./ui_shipper');
} else {
  import('./server_shipper');
}
```

So, essentially, the server-side will never be used on the UI and should not be included in the bundle. I wish Webpack were able to identify this and only import the parts of the code that are used. Otherwise, we are bundling entire libraries like `lodash` or `rxjs` only to use a small subset of methods. That seems a huge cost in terms of bundle size.

We can split the package in 3 to have `common`, `ui` and `server`. And it'll play nice for our package and monorepo approach in Kibana. However, given that this package will eventually make it to its own repo and NPM module, it feels like we'd be setting ourselves, the Platform Analytics team as the future maintainers of this library, and any other Elastic products using this client to a larger dependency tree. I wonder if we can do anything at the Bazel level to not expose everything at the root level so shippers can be imported on demand by the app without requiring the entire library? This could reduce `core` bundle size as it doesn't care about the shippers, as well as reducing `cloud` and `telemetry` plugins as they'd only import the shipper they are configuring (`fullstory` and `elastic_v3` respectively).
Exposing `mocks` from the same library is another use case for this that we now need to overcome by creating a separate package.

A potential different approach is to create a package for the client and a different one for each shipper. However, these are intended to be Elastic-approved shippers that _should_ be used when using the client. Having to install a separate library for each one of them feels like making them "optional".

I hope anything in my rambling makes sense. Please, let me know what's the best approach here, bearing in mind that it'll be eventually moved to a separate repo.

Reply:
I understand the idea, and it's something we've considered a lot in the design of the packaging system. The primary issue with supporting something like this is that it makes the build process for packages require a lot more customization which makes it more fragile and harder to reason about.

Right now every package implements it's own build process with a custom BUILD.bazel file and they are primarily written and maintained by Operations. Once we incorporate all code in the repo into package we can't have hundreds of bespoke build processes in use. Instead we've settled on a base set of build procedures which can be reused across packages. Supporting some level of nesting in these build procedures is technically possible, but complicated, and we've decided that it's not worth implementing because the simpler approach has an additional benefit: it helps discourage dependency bottlenecks in the system.

Dependency bottlenecks are like `lodash`, `rxjs`, or the `data` plugin where a single package is created to contain a bunch of separate things and consumers regularly depend on that package for a small fraction of it's API. These bottlenecks lead to circular dependencies, unnecessary rebuilds and test runs, cache thrashing, and many other issues. Their primary benefit is that there are less packages overall, but they have caused so many problems for us over the years and we're trying to get rid of them going forward. In the package system we're building we want to encourage folks to instead break those packages up into pieces which means more packages, enforcing the need to make it trivial to create packages and to standardize on the build procedures.

Maybe once we get past this initial rollout we can look at making more complicated package types, but for now we see a lot of simplicity in this approach and think the trade-offs are worth the added cost of having more packages. Please try the package generator:

```
node scripts/generate package --help
```

Creating new packages shouldn't be hard and will get even easier soon.

I hope this helps explain why we think this is the right way forward, if `@kbn/analytics` moves out of the Kibana repo then an entire build procedure for that new repo will have to be defined and I think it will be a perfect time to revisit the idea of using separate packages. I'd also be happy to help with ESLint rules at that time which automatically rewrites all the `@kbn/analytics-client` and `@kbn/analytics-server` imports.

Reply:
Makes sense! Thanks for explaining!

> node scripts/generate package --help

I actually used it to generate the `@elastic/analytics` package (only that I manually renamed it to `@elastic/` instead of `@kbn/' because of the already mentioned intention to eventually move it outside Kibana. 😇 It worked like a charm 😍 

I'll see how I can do to minimize the amount of code duplication while keeping a clear split between UI and shippers. I think the best approach can be:
1. Keep `@elastic/analytics` to hold the main client and types. This package is imported by `core` only.
2. Create `@elastic/analytics-shipper-fullstory` that imports the types only from `@elastic/analytics` and implements the FullStory shipper. This package is used by the `cloud` plugin.
3. Create `@elastic/analytics-shipper-elastic_v3-common` to hold the common logic and helpers for the "elastic V3" shipper.
4. Create `@elastic/analytics-shipper-elastic_v3-ui` that imports the `*-common` and includes the ui-side implementation of the shipper. Imported by the `telemetry` plugin in the `public` directory.
5. Create `@elastic/analytics-shipper-elastic_v3-server` that imports the `*-common` and includes the server-side implementation of the shipper. Imported by the `telemetry` plugin in the `server` directory.

With this split, we might not need to import add the package to the shared libs: https://github.com/elastic/kibana/issues/130027

Reply:
Nice! Sounds like you've got a good handle on how you want to lay this out and being able to avoid adding things to shared libs sounds like a win!